src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
-                             fact-ids)
-            (sql/do-prepared (format "DELETE FROM fact_paths fp
-                                      WHERE fp.id in ( SELECT fp.id
-                                                       FROM fact_paths fp left join fact_values fv on fp.id = fv.path_id
-                                                       WHERE fv.path_id is NULL )"))))))))
-
-(pls/defn-validated update-facts!
+  (sql/transaction
+   (sql/insert-record :factsets
+                      {:certname name
+                       :timestamp (to-timestamp timestamp)
+                       :environment_id (ensure-environment environment)
+                       :producer_timestamp (to-timestamp producer-timestamp)})
+   ;; Ensure that all the required paths and values exist, and then
+   ;; insert the new facts.
+   (let [paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+         pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+         valuemaps (map second paths-and-valuemaps)
+         vhashes (map :value_hash valuemaps)
+         path-ids (map (realize-paths! pathstrs) pathstrs)
+         vhash-ids (map :id (map (realize-values! valuemaps) vhashes))
+         pv-pairs (map vector path-ids vhash-ids)]
+     (insert-facts-pv-pairs! (certname-to-factset-id name) pv-pairs))))
+
+(defn-validated update-facts!
   "Given a certname, querys the DB for existing facts for that
    certname and will update, delete or insert the facts as necessary
-   to match the facts argument."
-  [{:keys [name values environment timestamp producer-timestamp] :as fact-data} :- facts-schema]
-  (let [factset (certname-to-factset-id name)
-        old-facts (current-fact-value-ids factset)
-        new-facts (new-fact-value-ids values)]
-    (sql/update-values :factsets ["id=?" factset]
-                       {:timestamp (to-timestamp timestamp)
-                        :environment_id (ensure-environment environment)
-                        :producer_timestamp (to-timestamp producer-timestamp)})
-    (utils/diff-fn (zipmap new-facts (repeat nil))
-                   (zipmap old-facts (repeat nil))
-                   #(insert-facts! factset %)
-                   #(delete-facts! factset %)
-                   identity)))
+   to match the facts argument. (cf. add-facts!)"
+  [{:keys [name values environment timestamp producer-timestamp] :as fact-data}]
+
+  ;; Might be possible to make some of this faster via sorted parallel traversal.
+  (sql/transaction
+   (let [factset-id (certname-to-factset-id name)
+         initial-factset-paths-vhashes
+         (sql/with-query-results result-set
+           ["SELECT fp.path, fv.value_hash FROM facts f
+               INNER JOIN fact_paths fp ON f.fact_path_id = fp.id
+               INNER JOIN fact_values fv ON f.fact_value_id = fv.id
+               WHERE factset_id = ?"
+            factset-id]
+           (doall result-set))
```

Comment:
@senior no reason, that was just before I'd gotten used to query-to-vec (fixed in upcoming push).